### PR TITLE
MGMT-16521: Identify Operator-based Deployment Data

### PR DESCRIPTION
**Pull request by CrystalChun**
https://issues.redhat.com/browse/MGMT-16521
Read the deployment data from the environment
variables when deploying with the operator.
Possible values include ACM, MCE, and operator
(indicates stand-alone operator).

The block querying for ACM and MCE deployments
are a temporary work-around until ACM/MCE
allows injecting environment variables into the
infrastructure operator deployment.

https://issues.redhat.com/browse/ACM-9362
Slack thread context https://redhat-internal.slack.com/archives/CUEMEHRA9/p1704996574257619

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

Test steps:

1. Built operator image with changes
2. Deploy ACM in OCP 4.14 aws cluster (from cluster bot)
3. Paused MCE and used operator image in infrastructure operator deployment
    ````
    $ oc annotate multiclusterengine multiclusterengine pause=true
    $ oc edit deploy infrastructure-operator -n multicluster-engine # change all the assisted-service images to custom image
    ````
5. Created AgentServiceConfig
6. Get the configmap in the multicluster engine namespace and verified correct values
    ````
    $ oc get -oyaml configmap assisted-service -n multicluster-engine | grep DEPLOYMENT
      DEPLOYMENT_TYPE: ACM
      DEPLOYMENT_VERSION: 2.9.2
    ````

Following the same steps, but deploying MCE stand-alone:
```
$ oc get -oyaml configmap assisted-service -n multicluster-engine | grep DEPLOYMENT
  DEPLOYMENT_TYPE: MCE
  DEPLOYMENT_VERSION: 2.3.4
```

Deploying the stand-alone operator:
1. Edit CSV to point to image with these changes ([here](https://github.com/openshift/assisted-service/blob/f5fc19587c15870b3dd03ad8060a6f1b70634c4f/deploy/olm-catalog/manifests/assisted-service-operator.clusterserviceversion.yaml#L965))
2. Build operator bundle
3. Create `assisted-installer` namespace and create quay pull secret in namespace
4. Deploy using operator-sdk to openshift 4.14 cluster
    ```
    $ operator-sdk run bundle \                             
      --namespace assisted-installer \
      ${BUNDLE_IMAGE:-quay.io/edge-infrastructure/assisted-service-operator-bundle:latest} --kubeconfig kubeconfig --pull-secret-name pull-secret
    ```
5. Create agentserviceconfig
6. Verify the configmap contains the correct values
    ```
    $ oc get configmap assisted-service -oyaml | grep DEPLOY
      DEPLOY_TARGET: k8s
      DEPLOYMENT_TYPE: Operator
      DEPLOYMENT_VERSION: quay.io/cchun/assisted-service:test-new-env-var-2
    ```

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 1]
```diff
@@ -1114,12 +1114,59 @@ func newAssistedCM(ctx context.Context, log logrus.FieldLogger, asc ASC) (client
...
 
+func getDeploymentData(ctx context.Context, cm *corev1.ConfigMap, asc ASC) {
+	if cm.Data["DEPLOYMENT_TYPE"] != "" && cm.Data["DEOLOYMENT_VERSION"] != "" {
+		return
+	}
+	//TODO: Remove this whole block and getVersionFromDeployment() function
+	// once ACM/MCE allows env var injection https://issues.redhat.com/browse/ACM-9362
+	deploy := &appsv1.Deployment{}
+	// Try getting the ACM deployment first
+	if err := asc.Client.Get(ctx, types.NamespacedName{Name: "multiclusterhub-operator", Namespace: "open-cluster-management"}, deploy); err == nil {
```
Comment (rccrdpccl) [inline internal/controller/controllers/agentserviceconfig_controller.go:1134]: What's the RBAC policy used for the running application? Did we try this in a real world scenario? There might be some RBAC limitations on fetching objects in other namespaces

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 2]
```diff
@@ -1114,12 +1114,57 @@ func newAssistedCM(ctx context.Context, log logrus.FieldLogger, asc ASC) (client
...
+	cm.Data["DEPLOYMENT_VERSION"] = ServiceImage()
+}
+
+// Extracts the environment variable OPERATOR_VERSION from a k8s deployment
+// Used for ACM/MCE deployments
+func getVersionFromDeployment(deploy *appsv1.Deployment, containerName string) string {
+	for _, container := range deploy.Spec.Template.Spec.Containers {
+		if container.Name == containerName {
+			for _, env := range container.Env {
+				if env.Name == "OPERATOR_VERSION" {
```
[the remark this answers is not in the document]
Comment (CrystalChun) [reply internal/controller/controllers/agentserviceconfig_controller.go:1168]: Spoke offline and realized that the unknown value is already sent. Unit tests have been added. Thanks!

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 3]
```diff
@@ -1114,12 +1114,57 @@ func newAssistedCM(ctx context.Context, log logrus.FieldLogger, asc ASC) (client
...
+	deploy := &appsv1.Deployment{}
+	// Try getting the ACM deployment first
+	if err := asc.Client.Get(ctx, types.NamespacedName{Name: "multiclusterhub-operator", Namespace: "open-cluster-management"}, deploy); err == nil {
+		cm.Data["DEPLOYMENT_TYPE"] = "ACM"
+		cm.Data["DEPLOYMENT_VERSION"] = getVersionFromDeployment(deploy, "multiclusterhub-operator")
+		return
+
+	}
+	// Try getting the MCE deployment if the ACM deployment does not exist
+	if err := asc.Client.Get(ctx, types.NamespacedName{Name: "multicluster-engine-operator", Namespace: "multicluster-engine"}, deploy); err == nil {
```
[the remark this answers is not in the document]
Comment (CrystalChun) [reply internal/controller/controllers/agentserviceconfig_controller.go]: Changed to constants. Thanks!

File: internal/controller/controllers/agentserviceconfig_controller.go (hunk 1 above)
Comment (CrystalChun) [reply internal/controller/controllers/agentserviceconfig_controller.go]: Good catch, I did test this out in a real world scenario (deployed OCP cluster using cluster bot then installed ACM via the operator hub) and it was able to query for the deployment alright, but I can look into RBAC generation to verify.

Comment (CrystalChun) [reply internal/controller/controllers/agentserviceconfig_controller.go]: Just checked and ACM does give us a clusterrole which allows us to see deployments in all namespaces
https://github.com/stolostron/backplane-operator/blob/57e760d08c6598915fc169d5c836e0f11c3af037/pkg/templates/charts/toggle/assisted-service/templates/assisted-service-clusterrole.yaml#L283-L286

**Review by rccrdpccl — APPROVED** [review]

**Comment by rccrdpccl** [comment]
/hold
probably needs to be merged after https://github.com/openshift/assisted-service/pull/5874

**Pull request merged**